File: backend/src/modules/transportation/exceptions/exceptions.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../../prisma/prisma.service';
import { fail, paginationHeader, success, successList, successWrite } from '../../../common/response/base-response';
import { ExceptionBody } from './exceptions.controller';

// Joins an HrUser name from its parts.
function fullName(u?: { firstName?: string | null; middleName?: string | null; lastName?: string | null } | null): string {
  if (!u) return '';
  return [u.firstName, u.middleName, u.lastName].filter(Boolean).join(' ');
}

// Rich include so list rows can be flattened to the frozen JSON keys.
const exceptionInclude = {
  hrUser: true,
  route: { include: { line: true } },
} as const;

@Injectable()
export class ExceptionsService {
  constructor(private readonly prisma: PrismaService) {}

  /** GET GetEmployeeException — paginated exceptions (TransportationExceptionData rows). */
  async getAll(pageNo = 1, noOfItems = 20, filters: { hrUserId?: number; id?: number } = {}) {
    const where: Record<string, unknown> = { active: true };
    if (filters.hrUserId) where.hrUserId = filters.hrUserId;
    if (filters.id) where.id = filters.id;

    const [total, rows] = await this.prisma.$transaction([
      this.prisma.transportationVehicleRouteEmployeeException.count({ where }),
      this.prisma.transportationVehicleRouteEmployeeException.findMany({
        where,
        include: exceptionInclude,
        orderBy: { id: 'desc' },
        skip: (pageNo - 1) * noOfItems,
        take: noOfItems,
      }),
    ]);

    // Resolve creator (User) names and direction names — neither has a direct
    // relation on the exception model, so batch-fetch both by id.
    const creatorMap = new Map<number, string>();
    const creatorIds = [...new Set(rows.map((e) => e.creationBy).filter((id): id is number => !!id))];
    if (creatorIds.length) {
      const users = await this.prisma.user.findMany({
        where: { id: { in: creatorIds } },
        select: { id: true, firstName: true, middleName: true, lastName: true, email: true },
      });
      for (const u of users) creatorMap.set(u.id, fullName(u) || u.email);
    }
    const directionMap = new Map<number, string>();
    const directionIds = [...new Set(rows.map((e) => e.directionId).filter((id): id is number => !!id))];
    if (directionIds.length) {
      const dirs = await this.prisma.transportationVehicleRouteDirection.findMany({
        where: { id: { in: directionIds } },
        select: { id: true, routeDirection: true },
      });
      for (const d of dirs) directionMap.set(d.id, d.routeDirection);
    }

    const data = rows.map((e) => ({
      Id: String(e.id),
      HrUserId: String(e.hrUserId),
      TransportationVehicleRouteId: e.routeId ?? 0,
      TransportationLineName: e.route?.line?.lineName ?? '',
      Active: e.active,
      CreationDate: e.creationDate ? e.creationDate.toISOString() : '',
      CreationBy: e.creationBy != null ? String(e.creationBy) : '',
      CreationName: e.creationBy != null ? (creatorMap.get(e.creationBy) ?? '') : '', // creator (User) full name
      TransportationVehicleRouteDirectionId: e.directionId != null ? String(e.directionId) : '',
      TransportationVehicleRouteDirectionName: e.directionId != null ? (directionMap.get(e.directionId) ?? '') : '', // direction/station name
      FromDate: e.fromDate ? e.fromDate.toISOString() : '',
      ToDate: e.toDate ? e.toDate.toISOString() : '',
      Period: e.period ?? '',
      DayName: e.dayName ?? '',
      LatitudeExceptional: 0, // TODO: no separate exceptional latitude field
      LongtitudExceptional: 0, // TODO: no separate exceptional longitude field (frozen typo key)
      ExceptionDate: e.exceptionDate ? e.exceptionDate.toISOString() : '',
      ExceptionDatePeriod: '', // TODO
      Latitude: e.latitude ?? 0,
      Longtitud: e.longitude ?? 0, // frozen typo key
      ExceptionDirectionId: '', // TODO: no separate exception direction field
      ExceptionDirectionName: '', // TODO
      HrUserName: fullName(e.hrUser),
      ReasonException: e.reasonException ?? '',
      ContactNumber: e.contactNumber ?? '',
    }));

    return successList(data, paginationHeader(pageNo, noOfItems, total));
  }

  /** POST AddException — create an exception. creationBy = caller. */
  async add(body: ExceptionBody, userId: number) {
    const routeId = Number(body?.transportationVehicleRouteId) || 0;
    const hrUserId = Number(body?.hrUserId) || 0;
    if (!routeId) return fail('Err101', 'transportationVehicleRouteId is required');
    if (!hrUserId) return fail('Err101', 'hrUserId is required');

    const created = await this.prisma.transportationVehicleRouteEmployeeException.create({
      data: this.toData(body, hrUserId, routeId, { creationBy: userId }),
    });
    return successWrite(created.id);
  }

  /** POST UpdateException — update an exception. */
  async update(id: number, body: ExceptionBody, userId: number) {
    if (!id) return fail('Err101', 'id is required');
    const routeId = Number(body?.transportationVehicleRouteId) || 0;
    const hrUserId = Number(body?.hrUserId) || 0;
    await this.prisma.transportationVehicleRouteEmployeeException.update({
      where: { id },
      data: this.toData(body, hrUserId, routeId, { creationBy: userId }),
    });
    return successWrite(id);
  }

  /**
   * GET GetCapacityNumbers — capacity breakdown for a route on a given day
   * (TransportationCapacityData). Faithful to the old GetCapacityNumbers
   * (svc:5642-5717): counts the route's scheduled riders for the day, then
   *   ActualCapacity = CapacityWithoutExpection + ExpectionNumFromOtherLines
   *                    - RouteEmployeesToOtherLines
   * where exceptions moving a rider off this route subtract, and exceptions
   * bringing a rider onto this route add.
   */
  async getCapacity(routeId: number, period?: string, dateSearch?: string) {
    if (!routeId) return fail('Err101', 'RouteID is required');
    const route = await this.prisma.transportationVehicleRoute.findUnique({
      where: { id: routeId },
      include: { vehicle: true },
    });
    if (!route) return fail('Err404', 'Route not found');

    // Evaluation day (old defaulted to "now" in Egypt time). Bound to the day.
    const target = dateSearch ? new Date(dateSearch) : new Date();
    const dayStart = new Date(Date.UTC(target.getUTCFullYear(), target.getUTCMonth(), target.getUTCDate()));
    const dayEnd = new Date(Date.UTC(target.getUTCFullYear(), target.getUTCMonth(), target.getUTCDate(), 23, 59, 59, 999));
    const dayName = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'][target.getUTCDay()];
    const periodValues = period ? [period, 'Both'] : ['Both'];

    // Riders scheduled on THIS route for the day: active membership, period
    // matches (or Both), and the day falls inside the membership window (when set).
    const members = await this.prisma.transportationVehicleRouteEmployee.findMany({
      where: {
        routeId,
        active: true,
        period: { in: periodValues },
        AND: [
          { OR: [{ fromDate: null }, { fromDate: { lte: dayEnd } }] },
          { OR: [{ toDate: null }, { toDate: { gte: dayStart } }] },
        ],
      },
      select: { hrUserId: true },
    });

    // Matching exceptions for that exact period on that day (old requires an
    // exact Period, not Both): either a date-range+weekday rule, or a one-off
    // ExceptionDate. Without a period we can't match, so none apply.
    const exceptions = period
      ? await this.prisma.transportationVehicleRouteEmployeeException.findMany({
          where: {
            active: true,
            OR: [
              {
                period,
                dayName,
                AND: [
                  { OR: [{ fromDate: null }, { fromDate: { lte: dayEnd } }] },
                  { OR: [{ toDate: null }, { toDate: { gte: dayStart } }] },
                ],
              },
              { period, exceptionDate: { gte: dayStart, lte: dayEnd } },
            ],
          },
          select: { hrUserId: true, routeId: true },
        })
      : [];

    const memberIds = new Set(members.map((m) => m.hrUserId));
    const exceptedUserIds = new Set(exceptions.map((e) => e.hrUserId));
    // This route's riders who have a matching exception (moved off the route).
    const routeEmployeesToOtherLines = [...memberIds].filter((id) => exceptedUserIds.has(id)).length;
    // Exceptions that target this route (riders brought onto the route).
    const expectionNumFromOtherLines = exceptions.filter((e) => e.routeId === routeId).length;
    const capacityWithoutException = members.length;

    const data = {
      FullCapacity: route.vehicle?.capacity ?? 0,
      ActualCapacity: capacityWithoutException + expectionNumFromOtherLines - routeEmployeesToOtherLines,
      CapacityWithoutExpection: capacityWithoutException, // frozen typo key
      ExpectionNumFromOtherLines: expectionNumFromOtherLines, // frozen typo key
      RouteEmployeesToOtherLines: routeEmployeesToOtherLines,
    };
    return success(data);
  }

  /** Maps the documented payload to the exception model columns. */
  private toData(body: ExceptionBody, hrUserId: number, routeId: number, extra: { creationBy: number }) {
    return {
      hrUserId,
      routeId,
      period: body.Period ?? '',
      fromDate: body.FromDate ? new Date(body.FromDate) : null,
      toDate: body.ToDate ? new Date(body.ToDate) : null,
      dayName: body.DayName ?? null,
      exceptionDate: body.ExceptionDate ? new Date(body.ExceptionDate) : null,
      latitude: body.Latitude != null ? Number(body.Latitude) : null,
      longitude: body.Longtitud != null ? Number(body.Longtitud) : null,
      directionId: body.TransportationVehicleRouteDirectionId
        ? Number(body.TransportationVehicleRouteDirectionId)
        : null,
      reasonException: body.reasonException ?? null,
      contactNumber: body.contactNumber ?? null,
      active: body.Active ?? true,
      creationBy: extra.creationBy,
    };
  }
}

File: backend-dotnet/GarasTransport.Api/Modules/Transportation/Exceptions/ExceptionsService.cs
using GarasTransport.Api.Common;
using Microsoft.EntityFrameworkCore;
using UserEntity = GarasTransport.Domain.Entities.User;

namespace GarasTransport.Api.Modules.Transportation.Exceptions;

// Shape of the Add/Update exception payload (documented keys, §6.8).
public class ExceptionBody
{
    public int? id { get; set; }
    public int? transportationVehicleRouteId { get; set; }
    public int? hrUserId { get; set; }
    public bool? Active { get; set; }
    public string? reasonException { get; set; }
    public string? contactNumber { get; set; }
    public string? Period { get; set; } // Go | Return | Both
    public string? FromDate { get; set; }
    public string? ToDate { get; set; }
    public string? DayName { get; set; }
    public string? ExceptionDate { get; set; }
    public double? Latitude { get; set; }
    public double? Longtitud { get; set; } // frozen typo key
    public int? TransportationVehicleRouteDirectionId { get; set; }
}

public class ExceptionsService
{
    private readonly AppDbContext _db;

    public ExceptionsService(AppDbContext db) => _db = db;

    private static string FullName(HrUser? u) =>
        u == null ? string.Empty : string.Join(' ', new[] { u.FirstName, u.MiddleName, u.LastName }.Where(s => !string.IsNullOrEmpty(s)));

    private static string FullName(UserEntity? u) =>
        u == null ? string.Empty : string.Join(' ', new[] { u.FirstName, u.MiddleName, u.LastName }.Where(s => !string.IsNullOrEmpty(s)));

    /// <summary>GET GetEmployeeException — paginated exceptions (TransportationExceptionData rows).</summary>
    public async Task<object> GetAll(int pageNo, int noOfItems, int? hrUserId, int? id)
    {
        var query = _db.TransportationVehicleRouteEmployeeExceptions.Where(e => e.Active);
        if (hrUserId.HasValue) query = query.Where(e => e.HrUserId == hrUserId.Value);
        if (id.HasValue) query = query.Where(e => e.Id == id.Value);

        var total = await query.CountAsync();
        var rows = await query
            .Include(e => e.HrUser)
            .Include(e => e.Route).ThenInclude(r => r!.Line)
            .OrderByDescending(e => e.Id)
            .Skip((pageNo - 1) * noOfItems)
            .Take(noOfItems)
            .ToListAsync();

        // Resolve creator (User) names and direction names — neither has a direct
        // relation on the exception model, so batch-fetch both by id.
        var creatorIds = rows.Where(e => e.CreationBy.HasValue).Select(e => e.CreationBy!.Value).Distinct().ToList();
        var creatorMap = new Dictionary<int, string>();
        if (creatorIds.Count > 0)
        {
            var users = await _db.Users.Where(u => creatorIds.Contains(u.Id)).ToListAsync();
            foreach (var u in users)
                creatorMap[u.Id] = string.IsNullOrEmpty(FullName(u)) ? u.Email : FullName(u);
        }
        var directionIds = rows.Where(e => e.DirectionId.HasValue).Select(e => e.DirectionId!.Value).Distinct().ToList();
        var directionMap = new Dictionary<int, string>();
        if (directionIds.Count > 0)
        {
            var dirs = await _db.TransportationVehicleRouteDirections.Where(d => directionIds.Contains(d.Id)).ToListAsync();
            foreach (var d in dirs) directionMap[d.Id] = d.RouteDirection;
        }

        var data = rows.Select(e => new
        {
            Id = e.Id.ToString(),
            HrUserId = e.HrUserId.ToString(),
            TransportationVehicleRouteId = e.RouteId,
            TransportationLineName = e.Route?.Line?.LineName ?? string.Empty,
            Active = e.Active,
            CreationDate = Fmt.Iso(e.CreationDate),
            CreationBy = e.CreationBy != null ? e.CreationBy.Value.ToString() : string.Empty,
            CreationName = e.CreationBy != null ? creatorMap.GetValueOrDefault(e.CreationBy.Value, string.Empty) : string.Empty,
            TransportationVehicleRouteDirectionId = e.DirectionId != null ? e.DirectionId.Value.ToString() : string.Empty,
            TransportationVehicleRouteDirectionName = e.DirectionId != null ? directionMap.GetValueOrDefault(e.DirectionId.Value, string.Empty) : string.Empty,
            FromDate = Fmt.Iso(e.FromDate),
            ToDate = Fmt.Iso(e.ToDate),
            Period = e.Period ?? string.Empty,
            DayName = e.DayName ?? string.Empty,
            LatitudeExceptional = 0,
            LongtitudExceptional = 0, // frozen typo key
            ExceptionDate = Fmt.Iso(e.ExceptionDate),
            ExceptionDatePeriod = string.Empty,
            Latitude = e.Latitude ?? 0,
            Longtitud = e.Longitude ?? 0, // frozen typo key
            ExceptionDirectionId = string.Empty,
            ExceptionDirectionName = string.Empty,
            HrUserName = FullName(e.HrUser),
            ReasonException = e.ReasonException ?? string.Empty,
            ContactNumber = e.ContactNumber ?? string.Empty,
        });

        return Envelope.SuccessList(data, Envelope.Pagination(pageNo, noOfItems, total));
    }

    /// <summary>POST AddException — create an exception. creationBy = caller.</summary>
    public async Task<object> Add(ExceptionBody body, int userId)
    {
        var routeId = body.transportationVehicleRouteId ?? 0;
        var hrUserId = body.hrUserId ?? 0;
        if (routeId == 0) return Envelope.Fail("Err101", "transportationVehicleRouteId is required");
        if (hrUserId == 0) return Envelope.Fail("Err101", "hrUserId is required");

        var entity = new TransportationVehicleRouteEmployeeException();
        Apply(entity, body, hrUserId, routeId, userId);
        _db.TransportationVehicleRouteEmployeeExceptions.Add(entity);
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(entity.Id);
    }

    /// <summary>POST UpdateException — update an exception.</summary>
    public async Task<object> Update(int id, ExceptionBody body, int userId)
    {
        if (id == 0) return Envelope.Fail("Err101", "id is required");
        var entity = await _db.TransportationVehicleRouteEmployeeExceptions.FirstOrDefaultAsync(e => e.Id == id);
        if (entity == null) return Envelope.Fail("Err404", "Exception not found");
        var routeId = body.transportationVehicleRouteId ?? 0;
        var hrUserId = body.hrUserId ?? 0;
        Apply(entity, body, hrUserId, routeId, userId);
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id);
    }

    /// <summary>GET GetCapacityNumbers — capacity breakdown for a route on a given day.</summary>
    public async Task<object> GetCapacity(int routeId, string? period, string? dateSearch)
    {
        if (routeId == 0) return Envelope.Fail("Err101", "RouteID is required");
        var route = await _db.TransportationVehicleRoutes.Include(r => r.Vehicle).FirstOrDefaultAsync(r => r.Id == routeId);
        if (route == null) return Envelope.Fail("Err404", "Route not found");

        var target = string.IsNullOrEmpty(dateSearch) ? DateTime.UtcNow
            : (DateTime.TryParse(dateSearch, null, System.Globalization.DateTimeStyles.AdjustToUniversal | System.Globalization.DateTimeStyles.AssumeUniversal, out var d) ? d : DateTime.UtcNow);
        var dayStart = new DateTime(target.Year, target.Month, target.Day, 0, 0, 0, DateTimeKind.Utc);
        var dayEnd = new DateTime(target.Year, target.Month, target.Day, 23, 59, 59, DateTimeKind.Utc).AddMilliseconds(999);
        string[] dayNames = { "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday" };
        var dayName = dayNames[(int)target.DayOfWeek];
        var periodValues = period != null ? new[] { period, "Both" } : new[] { "Both" };

        var members = await _db.TransportationVehicleRouteEmployees
            .Where(m => m.RouteId == routeId && m.Active && periodValues.Contains(m.Period)
                        && (m.FromDate == null || m.FromDate <= dayEnd)
                        && (m.ToDate == null || m.ToDate >= dayStart))
            .Select(m => m.HrUserId)
            .ToListAsync();

        List<(int HrUserId, int RouteId)> exceptions = new();
        if (period != null)
        {
            var ex = await _db.TransportationVehicleRouteEmployeeExceptions
                .Where(e => e.Active &&
                    ((e.Period == period && e.DayName == dayName
                        && (e.FromDate == null || e.FromDate <= dayEnd)
                        && (e.ToDate == null || e.ToDate >= dayStart))
                     || (e.Period == period && e.ExceptionDate >= dayStart && e.ExceptionDate <= dayEnd)))
                .Select(e => new { e.HrUserId, e.RouteId })
                .ToListAsync();
            exceptions = ex.Select(e => (e.HrUserId, e.RouteId)).ToList();
        }

        var memberIds = members.ToHashSet();
        var exceptedUserIds = exceptions.Select(e => e.HrUserId).ToHashSet();
        var routeEmployeesToOtherLines = memberIds.Count(id => exceptedUserIds.Contains(id));
        var expectionNumFromOtherLines = exceptions.Count(e => e.RouteId == routeId);
        var capacityWithoutException = members.Count;

        var data = new
        {
            FullCapacity = route.Vehicle?.Capacity ?? 0,
            ActualCapacity = capacityWithoutException + expectionNumFromOtherLines - routeEmployeesToOtherLines,
            CapacityWithoutExpection = capacityWithoutException, // frozen typo key
            ExpectionNumFromOtherLines = expectionNumFromOtherLines, // frozen typo key
            RouteEmployeesToOtherLines = routeEmployeesToOtherLines,
        };
        return Envelope.Success(data);
    }

    private static DateTime? ParseDate(string? s) =>
        string.IsNullOrEmpty(s) ? null : (DateTime.TryParse(s, null, System.Globalization.DateTimeStyles.AdjustToUniversal | System.Globalization.DateTimeStyles.AssumeUniversal, out var d) ? d : null);

    /// <summary>Maps the documented payload to the exception model columns.</summary>
    private static void Apply(TransportationVehicleRouteEmployeeException e, ExceptionBody body, int hrUserId, int routeId, int userId)
    {
        e.HrUserId = hrUserId;
        e.RouteId = routeId;
        e.Period = body.Period ?? string.Empty;
        e.FromDate = ParseDate(body.FromDate);
        e.ToDate = ParseDate(body.ToDate);
        e.DayName = body.DayName;
        e.ExceptionDate = ParseDate(body.ExceptionDate);
        e.Latitude = body.Latitude;
        e.Longitude = body.Longtitud;
        e.DirectionId = body.TransportationVehicleRouteDirectionId;
        e.ReasonException = body.reasonException;
        e.ContactNumber = body.contactNumber;
        e.Active = body.Active ?? true;
        e.CreationBy = userId;
    }
}
